src/telemetry/stats.c

```diff
@@ -338,27 +337,25 @@ static bool
 get_chunk_compression_stats(StatsContext *statsctx, const Chunk *chunk,
 							Form_compression_chunk_size compr_stats)
 {
+	TupleInfo *ti;
+	MemoryContext oldmcxt;
+
 	if (!ts_chunk_is_compressed(chunk))
 		return false;
 
+	/* Need to execute the scan functions on the long-lived memory context */
+	oldmcxt = MemoryContextSwitchTo(statsctx->compressed_chunk_stats_iterator.scankey_mcxt);
 	ts_scan_iterator_scan_key_reset(&statsctx->compressed_chunk_stats_iterator);
 	ts_scan_iterator_scan_key_init(&statsctx->compressed_chunk_stats_iterator,
 								   Anum_compression_chunk_size_pkey_chunk_id,
 								   BTEqualStrategyNumber,
 								   F_INT4EQ,
 								   Int32GetDatum(chunk->fd.id));
+	ts_scan_iterator_start_or_restart_scan(&statsctx->compressed_chunk_stats_iterator);
+	ti = ts_scan_iterator_next(&statsctx->compressed_chunk_stats_iterator);
+	MemoryContextSwitchTo(oldmcxt);
```

Comment:
Not sure I follow.

If the `scankey_mcxt` is the memory context used while scanning the table using the iterator, shouldn't these functions (`ts_scan_iterator_scan_key_reset`, `ts_scan_iterator_scan_key_init`, etc.) *always* allocate any memory in this context? Why should it be necessary to switch to this context before calling the functions?

Or put another way, shouldn't this memory context switch be inside each function rather than requiring the user of those function realize that it is necessary to switch to a memory context owned by the iterator?

Reply:
Yes, but I didn't want to refactor in this PR. Instead, I wanted fix what is already released.

Reply:
Does it break if you switch context inside all the scan iterator functions? To me, this seems like the safest approach.

Do you know if there are any other locations with similar issues?

Reply:
I don't think there are other locations because we generally don't use "per tuple memory contexts"when scanning and instead just scan things on the main transaction memory context. In fact, these other locations are the reason I didn't want to change too much inside the scan functions as that would affect all other locations. "Don't fix what ain't broken".

In this code, I tried to be vigilant about not eating a lot of memory for these scans as they often involve an unbounded number of chunks. But I guess I didn't fully realize that most PG scan functions allocate new memory and don't use their own memory context internally.

It might make sense to wrap these APIs to use internal memory contexts that makes it easier to use per-tuple contexts,  as you suggest, but I didn't think that kind of refactoring was appropriate for this "fix" PR. 

If you feel it is important that we do it now, I can do it.

Reply:
I added a commit with the changes you suggested. But I need to ensure it passes Sqlsmith.

Reply:
I was concerned about that this fix does not address yet undiscovered issues that are lingering, not that I think the refactoring is necessary. It just seemed like the simplest and safest way to fix the issue.